node_groups.tf

```diff
@@ -1,112 +1,29 @@
-  count      = var.create_eks && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
-  role       = aws_iam_role.node_groups[0].name
-}
-
-resource "aws_iam_role_policy_attachment" "node_groups_additional_policies" {
-  for_each = var.create_eks && local.worker_group_managed_node_group_count > 0 ? toset(var.workers_additional_policies) : []
-
-  role       = aws_iam_role.node_groups[0].name
-  policy_arn = each.key
-}
-
-resource "aws_iam_role_policy_attachment" "node_groups_autoscaling" {
-  count      = var.create_eks && var.manage_worker_autoscaling_policy && var.attach_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  policy_arn = aws_iam_policy.node_groups_autoscaling[0].arn
-  role       = aws_iam_role.node_groups[0].name
-}
-
-resource "aws_iam_policy" "node_groups_autoscaling" {
-  count       = var.create_eks && var.manage_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  name_prefix = "eks-worker-autoscaling-${aws_eks_cluster.this[0].name}"
-  description = "EKS worker node autoscaling policy for cluster ${aws_eks_cluster.this[0].name}"
-  policy      = data.aws_iam_policy_document.worker_autoscaling[0].json
-  path        = var.iam_path
-}
-
-resource "random_pet" "node_groups" {
-  for_each = var.create_eks ? local.node_groups : {}
-
-  separator = "-"
-  length    = 2
-
-  keepers = {
-    instance_type = lookup(each.value, "instance_type", local.workers_group_defaults["instance_type"])
-
-    ec2_ssh_key = lookup(each.value, "key_name", local.workers_group_defaults["key_name"])
-
-    source_security_group_ids = join("-", compact(
-      lookup(each.value, "source_security_group_ids", local.workers_group_defaults["source_security_group_id"]
-    )))
-
-    node_group_name = join("-", [var.cluster_name, each.value["name"]])
+# Hack to ensure ordering of resource creation. Do not create node_groups
+# before other resources are ready. Removes race conditions
+data "null_data_source" "node_groups" {
+  count = var.create_eks ? 1 : 0
+
+  inputs = {
+    cluster_name = var.cluster_name
+
+    # Ensure these resources are created before "unlocking" the data source.
+    # `depends_on` causes a refresh on every run so is useless here.
+    # [Re]creating or removing these resources will trigger recreation of Node Group resources
+    aws_auth         = coalescelist(kubernetes_config_map.aws_auth[*].id, [""])[0]
+    role_NodePolicy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKSWorkerNodePolicy[*].id, [""])[0]
+    role_CNI_Policy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKS_CNI_Policy[*].id, [""])[0]
+    role_Container   = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEC2ContainerRegistryReadOnly[*].id, [""])[0]
+    role_auotscaling = coalescelist(aws_iam_role_policy_attachment.workers_autoscaling[*].id, [""])[0]
```

Comment:
Typo: should be role_autoscaling I guess

Reply:
Fixed, thank you.